js/src/components/supported-country-select/supported-country-select.js

```diff
@@ -0,0 +1,46 @@
+/**
```

Comment:
❓  Whats the reason of using 

`js/src/components/supported-country-select/index.js`  for exporting `js/src/components/supported-country-select/supported-country-select.js` and not just directly use `js/src/components/supported-country-select/index.js` or `js/src/components/supported-country-select/supported-country-select.js` ?

Maybe I missed some discussion for the convention?

Reply:
If I remember right, this was advocated by @tomalec. I got this convention (or preference?) during one of code reviews a few months ago, but I couldn't find that comment. The most relevant post I can find is p6q8Tx-1Uk-p2. I also lean toward this method, which brings more benefits when using the editor to find files, and when moving files via git.

Reply:
> not just directly use `js/src/components/supported-country-select/index.js`...

We have this in our code base, and we end up having a lot of code file with `index.js`, and it makes searching and navigating the files in the editors cumbersome, e.g. you will have a lot of "index.js" files in your tabs; when you do cmd+P in VS code and search for "index.js" a lot of results show up, etc. Personally it is not a good experience for me.

> or `js/src/components/supported-country-select/supported-country-select.js` ?

If we do this without the index.js, then when we import the component, the import statement would be longer, like:

```js
import SupportedCountrySelect from '.~/components/supported-country-select/supported-country-select';
```

By having `js/src/components/supported-country-select/index.js` for exporting `js/src/components/supported-country-select/supported-country-select.js`, then we would have a shorter import statement: 

```js
import SupportedCountrySelect from '.~/components/supported-country-select';
```

@eason9487 gave a good P2 post there in his comment. Another related good one is p90Yrv-2d4-p2.

Reply:
> when you do cmd+P in VS code and search for "index.js" a lot of results show up, etc. Personally it is not a good experience for me.

Same. It's inconvenient in Sublime Text too.

Reply:
I missed those post. 

> By having js/src/components/supported-country-select/index.js for exporting js/src/components/supported-country-select/supported-country-select.js, then we would have a shorter import statement:

Is the name long really important? I mean with IDE autocompletion I don't get why is bad to have long names. 

My thoughts are that if we are using `index.js` we still have the same problem right?

If you search for `supported-country-select` in this case. It will show you as well the index since the directory matching the search query as well.

Reply:
BTW. I agree with not using `index.js` as a name. I guess was also discussed in Google Dev Tools meeting.

Reply:
> Is the name long really important? I mean with IDE autocompletion I don't get why is bad to have long names.

I think it is not about "bad to have long names", it's just that it is "nicer to have shorter import statement". 😆 

Another thing about re-export in `index.js` in a folder is that it tells which files are the main things exported out from the folder, and which files are not meant to be used outside the folder. For example, in https://github.com/woocommerce/google-listings-and-ads/tree/develop/js/src/components/shipping-rate-section/estimated-shipping-rates-card, there are a number of files. By looking at `index.js`, it serves as a hint to us that only `EstimatedShippingRatesCard` are meant to be used outside the folder, while other files are meant to be used within the folder.

Reply:
> If you search for `supported-country-select` in this case. It will show you as well the index since the directory matching the search query as well.

Yes, you will get `index.js` as a result as well, but `supported-country-select.js` has a higher match :) Plus that's file name that is used/emphasized in many other places: search results, git status, debugger, etc.
![image](https://user-images.githubusercontent.com/17435/165378100-9aafa1fb-f79b-46c5-9b6e-50efe41048f7.png)

> Another thing about re-export in `index.js` in a folder is that it tells which files are the main things exported out from the folder, and which files are not meant to be used outside the folder […]

:+1: 
In my mental model, thanks to that approach folder structure may improve modular DevX. If a folder contains a module/component of any kind (as opposed to a directory of for example `utils/`) any file or folder inside is considered an implementation detail, and only things exported explicitly in `index.js` are meant/recommended to be used outside. For me that goes in pair with mapping `./my-module`: `./my-module.js`, `./my-module/index.js`

Reply:
> By looking at `index.js`, it serves as a hint to us that only `EstimatedShippingRatesCard` are meant to be used outside the folder, while other files are meant to be used within the folder.
> and only things exported explicitly in index.js are meant/recommended to be used outside.

I get the point, but still feels annoying for me the need of create `index.js` file for exporting the real entry point (estimated-shipping-rates-card.js), 2 files instead of 1 per each component created. 

In case is needed to add multiple exports then yes, would be good to use that `index.js`, besides that I liked more when exportable components are in separated and isolated folders. 

As an example, if I have a component in the directory FormField  with the component <FormField> that uses internally a <Label> component and that <Label> could be used "outside" IMO <Label> should be created outside FormField folder, hence no real need for that `index.js`.